r/R/expression.R

```diff
@@ -76,7 +76,15 @@
 Expression <- R6Class("Expression", inherit = ArrowObject,
   public = list(
     ToString = function() compute___expr__ToString(self),
-    type = function(schema) compute___expr__type(self, schema),
+    schema = NULL,
+    bind = function(schema) self$schema <- schema,
+    type = function() {
+      if (is.null(self$schema)) {
+        stop("Must bind() expression to a schema before returning its type", call. = FALSE)
+      }
+      compute___expr__type(self, self$schema)
```

Comment:
How about this?

```suggestion
    type = function(schema = self$schema) {
      assert_that(!is.null(schema))
      compute___expr__type(self, schema)
```

I think it's misleading to have a `$bind()` method that doesn't actually Bind the Expression. I think you can achieve the same by doing `expr$schema <- schema` inside `relocate()` And also this way you retain the schema arg to `$type()` (so you don't need to modify the print method above)

Reply:
Done in 285f82606786d4ff850dbacc515c59466caada6d

I also did a little test and found out that it's 2x as fast to get the type ID by calling a C++ function dedicated for that purpose compared to calling the existing type function and getting the ID from it, so I implemented it that way.

Reply:
Just for posterity, here's the test code I ran
```r
library(dplyr)
library(arrow)

x <- as.data.frame(t(c(1:26)))
colnames(x) <- letters
x$z <- "z"
x <- Table$create(x)

start_time <- Sys.time()
for (i in 1:100) {
  q <- x %>% relocate(where(is.numeric), .after = where(is.character))
  w <- x %>% relocate(where(is.character), .before = where(is.numeric))
}
end_time <- Sys.time()
end_time - start_time

# ~8 seconds before change
# ~4 seconds after change
```